src/hotspot/share/services/diagnosticCommand.cpp

```diff
@@ -1039,8 +1049,15 @@ void DumpSharedArchiveDCmd::execute(DCmdSource source, TRAPS) {
 
   // call CDS.dumpSharedArchive
   Handle fileh;
+  char fname[JVM_MAXPATHLEN];
   if (file != nullptr) {
-    fileh =  java_lang_String::create_from_str(_filename.value(), CHECK);
+    // If the filename contains %p, it will be replaced by the pid.
+    const char *name = _filename.value();
+    if (Arguments::copy_expand_pid(name, strlen(name), fname, JVM_MAXPATHLEN)) {
+      fileh = java_lang_String::create_from_str(fname, CHECK);
+    } else {
+      warning("Invalid file path name specified, fall back to default name");
```

Comment:
`warning` prints a warning to the stdout of the JVM process. You don't want that; you want a warning to the issuer of the dcmd, which is another - possibly even remote - process. Write errors to `output()`, instead.